src/main/java/org/jenkinsci/plugins/casc/EnvSecretSource.java

```diff
@@ -0,0 +1,23 @@
+package org.jenkinsci.plugins.casc;
+
+import hudson.Extension;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Extension
+public class EnvSecretSource extends SecretSource {
+
+    @Override
+    public String reveal(String fromKey) {
```

Comment:
`reveal` should only search to a matching variable in source, and the pattern matching code can move to SecretSource so it's used by all implementations.
Better return `Optional<String>` vs IllegalStateException when a specific SecretSource implementation can't find variable, only throw ex at higher level if none of the available SecretSource did found a match.